docs/source/advanced_usage.rst

```diff
@@ -35,21 +35,21 @@ locally):
 Concurrency
 ~~~~~~~~~~~
 
-Using pyQuil for concurrent programming is as simple as calling :py:func:`~pyquil.get_qc()` from within a given thread
-or process, then using the returned :py:class:`~pyquil.api.QuantumComputer` as usual. While
-:py:class:`~pyquil.api.QuantumComputer` objects as a whole are not safe to share between threads or processes (due to
-state related to currently-running compilation or execution requests), some information they use is. Information related
-to client configuration (:py:class:`~pyquil.api.QCSClientConfiguration`) and QPU auth
-(:py:class:`~pyquil.api.EngagementManager`) can be safely extracted and shared among
-:py:class:`~pyquil.api.QuantumComputer` instances, as shown below, to save your code from redundant disk reads and
-auth-related HTTP requests.
+:py:class:`~pyquil.api.QuantumComputer` objects are safe to share between threads or processes,
```

Comment:
Are you completely sure this is true? What about the statefulness of the underlying `AbstractCompiler` (and possibly `QAM`)? Even though `QuantumComputer` is far less stateful than before, there are a few areas to keep in mind (e.g. `QPUCompiler._calibration_program`, `QAM.experiment` [doesn't look used though], etc.).

Most of these look to be set in the constructor and not mutated after, so they're probably OK, but might be worth another pass to see if any of these, and others, are at risk for data races.

Reply:
True, while much of the state is set in the ctor and is OK in most cases, users could still choose to mutate them in threads, even if our APIs don't do that naturally. If we wanted true thread safety, we'd need to put every property behind getters and setters that can lock/unlock appropriately. Moreover, that only works for non-mutable types, like ints, strings, etc. If you had a mutable property that wasn't naturally thread-safe, locking around a getter/setter would do no good if the item itself is getting mutated unsafely (i.e. `qc.foo["x"] = 42`, vs `qc.foo = {...}`). Disclaimer: not sure if dict was a good example, as they may be thread-safe for some operations, but I'm sure you get the point I'm illustrating.